src/Features/Core/Portable/CodeStyle/AbstractCodeStyleProvider.Analysis.cs

```diff
@@ -0,0 +1,98 @@
+            if (analyzeSyntaxTree != null) { context.RegisterSyntaxTreeAction(c => AnalyzeIfEnabled(c, analyzeSyntaxTree)); }
+            if (analyzeSyntax != null) { context.RegisterSyntaxNodeAction(c => AnalyzeIfEnabled(c, analyzeSyntax), syntaxKinds); }
+            if (analyzeOperation != null) { context.RegisterOperationAction(c => AnalyzeIfEnabled(c, analyzeOperation), operationKinds); }
+
+            // check for all the other things that can be registered as well.
+        }
+
+        private void AnalyzeIfEnabled(CodeBlockAnalysisContext context, Action<CodeBlockAnalysisContext, CodeStyleOption<TOptionKind>> analyze)
+            => AnalyzeIfEnabled(context, analyze, context.Options, context.SemanticModel.SyntaxTree, context.CancellationToken);
+
+        private void AnalyzeIfEnabled(OperationAnalysisContext context, Action<OperationAnalysisContext, CodeStyleOption<TOptionKind>> analyze)
+            => AnalyzeIfEnabled(context, analyze, context.Options, context.Operation.SemanticModel.SyntaxTree, context.CancellationToken);
+
+        private void AnalyzeIfEnabled(SemanticModelAnalysisContext context, Action<SemanticModelAnalysisContext, CodeStyleOption<TOptionKind>> analyze)
+            => AnalyzeIfEnabled(context, analyze, context.Options, context.SemanticModel.SyntaxTree, context.CancellationToken);
+
+        private void AnalyzeIfEnabled(SyntaxNodeAnalysisContext context, Action<SyntaxNodeAnalysisContext, CodeStyleOption<TOptionKind>> analyze)
+            => AnalyzeIfEnabled(context, analyze, context.Options, context.SemanticModel.SyntaxTree, context.CancellationToken);
+
+        private void AnalyzeIfEnabled(SyntaxTreeAnalysisContext context, Action<SyntaxTreeAnalysisContext, CodeStyleOption<TOptionKind>> analyze)
+            => AnalyzeIfEnabled(context, analyze, context.Options, context.Tree, context.CancellationToken);
+
+        private void AnalyzeIfEnabled<TContext>(
+            TContext context, Action<TContext, CodeStyleOption<TOptionKind>> analyze,
+            AnalyzerOptions options, SyntaxTree syntaxTree, CancellationToken cancellationToken)
+        {
+            var optionSet = options.GetDocumentOptionSetAsync(syntaxTree, cancellationToken).GetAwaiter().GetResult();
+            if (optionSet == null)
+            {
+                return;
+            }
+
+            var optionValue = optionSet.GetOption(_option);
+            switch (optionValue.Notification.Severity)
+            {
+                case ReportDiagnostic.Error:
+                case ReportDiagnostic.Warn:
+                case ReportDiagnostic.Info:
+                    break;
+                default:
+                    // don't analyze if it's any other value.
+                    return;
+            }
+
+            analyze(context, optionValue);
+        }
+
+        public abstract class DiagnosticAnalyzer : AbstractCodeStyleDiagnosticAnalyzer
```

Comment:
here are the impls of hte actual DiagnosticAnalyzer/CodeRefactoringProvider/CodeFixProvider. They just defer everything to the AbstractCodeStyleProvider. Ideally they would be sealed types. 

However, because people have to export DiagnosticAnalyzer/CodeRefactoringProvider/CodeFixProvider themselves, these are exposed and abstract solely for the purpose of creating an exported subclass. You'll see that later on.